src/auth/googleauth.ts

```diff
@@ -615,4 +651,64 @@ export class GoogleAuth {
 
     return {client_email: data.default.email};
   }
+
+  /**
+   * Automatically obtain a client based on the provided configuration.  If no
+   * options were passed, use Application Default Credentials.
+   */
+  async getClient() {
+    if (!this.cachedCredential) {
+      if (this.keyfileName) {
+        const filePath = path.resolve(process.cwd(), this.keyfileName);
+        console.log(filePath);
+        const stream = fs.createReadStream(filePath);
+        this.cachedCredential = await this.fromStreamAsync(stream);
+      } else if (this.jsonContent) {
+        this.cachedCredential = await this.fromJSON(this.jsonContent);
+      } else {
+        await this.getApplicationDefaultAsync();
+      }
+    }
+    return this.cachedCredential!;
+  }
+
+  /**
+   * Automatically obtain application default credentials, and return
+   * an access token for making requests.
+   */
+  async getAccessToken() {
+    const client = await this.getClient();
+    return (await client.getAccessToken()).token;
+  }
+
+  /**
+   * Obtain the HTTP headers that will provide authorization for a given
+   * request.
+   */
+  async getRequestHeaders(url?: string) {
+    const client = await this.getClient();
+    return (await client.getRequestMetadata(url)).headers;
+  }
+
+  /**
+   * Obtain credentials for a request, then attach the appropriate headers to
+   * the request options.
+   * @param opts Axios or Request options on which to attach the headers
+   */
+  async authorizeRequest(
+      opts: {url?: string, uri?: string, headers?: http.IncomingHttpHeaders}) {
+    opts = opts || {};
+    const url = opts.url || opts.uri;
+    const client = await this.getClient();
+    const {headers} = await client.getRequestMetadata(url);
+    opts.headers = Object.assign(opts.headers || {}, headers);
+    return opts;
+  }
+
+  /**
+   * Determine the compute environment in which the code is running.
+   */
+  async getEnv(): Promise<GCPEnv> {
```

Comment:
Does this need to be `async`?

Reply:
Yup.  The checks for GCE/GKE are async since they contact the metadata server.

Reply:
I meant: you don't have an await and it is a single line function, so you can drop the `async` keyword and avoid some unnecessary promise wrapping.

Reply:
Ahh, I got it.  Removed.